Improve performance of `ProvideId` method

akopich opened the pull request:
The improvement consists in optimizing `ScopeTypeFromTimerInfo` function which is twofold:

1. The nested branching is replaced by a table and arithmetic operations
2. The function itself is turned into a lambda to facilitate inlining.

The speed-up is measured at the level of 15%.

Test: Unit, manual
Bug: http://b/264439413

karupayun commented on src/ClientData/include/ClientData/ScopeIdProvider.h:
````diff
@@ -107,8 +108,32 @@ class NameEqualityScopeIdProvider : public ScopeIdProvider {
+  //  }
+  //}
+  //```
+  static constexpr std::array<int, 14> kTable = {1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 3};
````
Consider explaining in a comment the meaning of each number (0 means `ScopeType::kInvalid`, ...)

karupayun commented on src/ClientData/include/ClientData/ScopeIdProvider.h:
```diff
@@ -107,8 +108,32 @@ class NameEqualityScopeIdProvider : public ScopeIdProvider {
+  // static ScopeType ScopeTypeFromTimerInfo(const TimerInfo& timer) {
+  //  switch (timer.type()) {
+  //    case TimerInfo::kNone:
+  //      return timer.function_id() != orbit_grpc_protos::kInvalidFunctionId
```
Since you modified the code, consider switching the if statement:

```
//      return timer.function_id() == orbit_grpc_protos::kInvalidFunctionId
//                 ? ScopeType::kInvalid;
//                 : ScopeType::kDynamicallyInstrumentedFunction
```

karupayun commented on src/ClientData/include/ClientData/ScopeIdProvider.h:
```diff
@@ -107,8 +108,32 @@ class NameEqualityScopeIdProvider : public ScopeIdProvider {
+  static constexpr auto kScopeTypeFromTimerInfo = [](const TimerInfo& timer) -> ScopeType {
+    const int type = timer.type();
+    const int index =
+        type ^ static_cast<int>(timer.function_id() == orbit_grpc_protos::kInvalidFunctionId);
```
This is not the same function than before. Please explain in a comment why this won't affect the other cases of the switch.

karupayun reviewed (commented):
In general looks good, but I'm not 100% sure if commenting the old code is the way to go, @beckerhe WDYT?

akopich commented:
I have added a comment. Also, I've fixed a bug, so the table has grown. 

Regarding the old implementation which I've commented out. If I was writing this from scratch, I'd probably write some pseudocode to document the function, so that the reader has an idea of what we're trying to achieve here. But hey -- we've already got plain C++ code doing just that.

akopich commented:
Converting to draft. I've got an idea how to make it even faster. I'll test the idea on Monday.

akopich commented:
Careful analysis reveled no statistically significant performance improvement. Closing.

akopich commented:
As pointer out by @beckerhe the performance issue we observe is probably due to cache incoherence.
